libsignal-service-dotnet/websocket/WebSocketWrapper.cs

```diff
@@ -29,36 +29,96 @@ public class WebSocketWrapper
 
         public WebSocketWrapper(string uri, CancellationToken token)
         {
-            _ws = new ClientWebSocket();
-            _ws.Options.KeepAliveInterval = TimeSpan.FromSeconds(20);
+            CreateSocket();
             _uri = new Uri(uri);
             Token = token;
         }
+        private void CreateSocket()
+        {
+            _ws = new ClientWebSocket();
+            _ws.Options.KeepAliveInterval = TimeSpan.FromSeconds(20);
+        }
 
         public void HandleOutgoingWS()
         {
             Debug.WriteLine(TAG + "HandleOutgoingWS started");
+            byte[] buf = null;
             while (!Token.IsCancellationRequested)
             {
+                
                 try
                 {
-                    var buf = OutgoingQueue.Take(Token);
+                    if (buf == null)
+                        buf = OutgoingQueue.Take(Token);
+                    ReconnectTask?.Wait();
                     _ws.SendAsync(new ArraySegment<byte>(buf, 0, buf.Length), WebSocketMessageType.Binary, true, Token).Wait();
+                    buf = null; //set to null so we do not retry the same block
                 }
                 catch (TaskCanceledException e)
                 {
+                    if (!Token.IsCancellationRequested)
+                        Reconnect();
                     Debug.WriteLine(TAG + "HandleOutgoingWS shutting down");
                 }
                 catch (Exception e)
                 {
                     Debug.WriteLine(TAG + "WS SendAsync failed: " + e.Message);
-                    //TODO reconnect
+                    Reconnect();
                 }
             }
             //TODO dispose
             Debug.WriteLine(TAG + "HandleOutgoingWS finished");
         }
+        private Task ReconnectTask;
+        public void Reconnect()
+        {
+            if (Token.IsCancellationRequested)
+                return;
+            if (ReconnectTask != null)
+            {
+                ReconnectTask?.Wait();//race condition handling
```

Comment:
This is unfortunately still racy, i think.

Imagine 2 callers calling `Reconnect()` and arriving at the construction of the TaskCompletionSource at the same time. They will both have passed the `ReconnectTask != null {return}` clause because at that time no `ReconnectTask` was set, and now they will both create a `TaskCompletionSource` and can simultaneously call `CloseOutputAsync`, `CreateSocket` and `ConnectAsync`.

For the sake of simplicity, can we just guard the interiors of `Reconnect()` with an old-school lock, read the socket [state ](https://msdn.microsoft.com/de-de/library/system.net.websockets.clientwebsocket.state(v=vs.110).aspx) at the beginning, and return if it is connected? That should be a foolproof way of ensuring that
- only one task is attempting to reconnect at the same time
- no task is wasting resources by trying to send on a non-open socket more than once without ensuring that a reconnect succeeded
- each thread will securely wait until the reconnect succeeds